Box-shadow repaint is broken in some writing modes

smfr opened the pull request:
#### d69f4715b015bb45ab9a045d3c778c592397edac
<pre>
Box-shadow repaint is broken in some writing modes
<a href="https://bugs.webkit.org/show_bug.cgi?id=294179">https://bugs.webkit.org/show_bug.cgi?id=294179</a>
<a href="https://rdar.apple.com/152803240">rdar://152803240</a>

Reviewed by Alan Baradlay.

Visual overflow is in the block&apos;s coordinate space (see the comment in `RenderBox::allowedLayoutOverflow()`),
so overflow from box-shadow (and the other contributors like border-image) need to handle that.

* LayoutTests/fast/box-shadow/shadow-box-resize-writing-mode-expected.txt: Progression; the test is also visibly progressed.
* LayoutTests/fast/repaint/box-shadow-repaint-sideways-writing-modes-expected.txt: Added.
* LayoutTests/fast/repaint/box-shadow-repaint-sideways-writing-modes.html: Added.
* LayoutTests/fast/repaint/box-shadow-repaint-writing-modes-expected.txt: Added.
* LayoutTests/fast/repaint/box-shadow-repaint-writing-modes.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::shadowOutsetsForWritingMode): Handle the confusing negative top/left that box-shadow outsets use,
and flip along the block axis if necessary.
(WebCore::RenderBox::applyVisualEffectOverflow const): Adjust the shadow offsets as necessary.
(WebCore::RenderBox::flippedClientBoxRect const): Drive-by cleanup.

Canonical link: <a href="https://commits.webkit.org/295977@main">https://commits.webkit.org/295977@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f76f148020548bada9ceff0f9c2651ac619754d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106822 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26573 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16971 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112033 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57420 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108861 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35075 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81115 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109826 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21628 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96396 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61456 "Build is in progress. Recent messages:Running run-api-tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21068 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14504 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56874 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90969 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14532 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114991 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33960 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25053 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90177 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34325 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92627 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89887 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22935 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34827 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12642 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29645 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33885 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39298 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33632 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36985 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35230 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/repaint/vertical-rl-box-shadow-repaint.html:
Should we add a test for sideways-lr / sideways-rl ? they sometimes need special treatment too.

weinig commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -4336,6 +4336,28 @@ void RenderBox::addVisualEffectOverflow()
         fragmentedFlow->addFragmentsVisualEffectOverflow(*this);
 }
 
+static LayoutBoxExtent shadowOutsetsForWritingMode(const LayoutBoxExtent& shadowExtent, WritingMode writingMode)
```
What about adding an overload to RenderStyle::boxShadowExtent() that takes a WritingMode instead? There have been more and more uses of that idiom going in and I think it makes sense for this as well.

Then the actual implementation would go StyleShadow.h and be usable for others as well.

alanbaradlay replied:
is that really a computed style thing? (though I guess RenderStyle is becoming less and less computed style but I still think it is the wrong direction)

smfr replied:
> What about adding an overload to RenderStyle::boxShadowExtent() that takes a WritingMode instead

In this case I think that's a bad idea. This code is specific to overflow rect computation, which uses an odd coordinate system.

weinig replied:
> is that really a computed style thing? (though I guess RenderStyle is becoming less and less computed style but I still think it is the wrong direction).

In that case, should we look at moving the other functions that take WritingModes? 

Honestly, the least computed style part of the thing to me is `RenderStyle::boxShadowExtent()`. It would probably better if that was instead just `Style::shadowOutsetExtent(renderStyle->boxShadow())`.

weinig replied:
> In this case I think that's a bad idea. This code is specific to overflow rect computation, which uses an odd coordinate system.

In that case, I think giving it a name that makes it clear that it would not be more generally useful would be good. Perhaps adding the words "OverflowRect" in the function name?

Though, I think perhaps a nicer composition would be:

```
    auto extent = Style::shadowOutsetExtentForWritingMode(renderStyle->boxShadow(), writingMode);
    auto extentInOverflowRectCoordinates = convertToOverflowRectCoordinates(extent);
```

making a weird coordinate system explicit.

smfr replied:
I'll address this in a followup.

nt1m commented on LayoutTests/fast/repaint/vertical-rl-box-shadow-repaint.html:
```diff
@@ -0,0 +1,42 @@
+
+    <div class="container" style="writing-mode: horizontal-bt">
+        <div class="box"></div>
+    </div>
```
i would love a test for sideways writing modes too if that's possible!

smfr replied:
Added.

smfr committed:
Box-shadow repaint is broken in some writing modes
https://bugs.webkit.org/show_bug.cgi?id=294179
rdar://152803240

Reviewed by Alan Baradlay.

Visual overflow is in the block's coordinate space (see the comment in `RenderBox::allowedLayoutOverflow()`),
so overflow from box-shadow (and the other contributors like border-image) need to handle that.

* LayoutTests/fast/box-shadow/shadow-box-resize-writing-mode-expected.txt: Progression; the test is also visibly progressed.
* LayoutTests/fast/repaint/box-shadow-repaint-sideways-writing-modes-expected.txt: Added.
* LayoutTests/fast/repaint/box-shadow-repaint-sideways-writing-modes.html: Added.
* LayoutTests/fast/repaint/box-shadow-repaint-writing-modes-expected.txt: Added.
* LayoutTests/fast/repaint/box-shadow-repaint-writing-modes.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::shadowOutsetsForWritingMode): Handle the confusing negative top/left that box-shadow outsets use,
and flip along the block axis if necessary.
(WebCore::RenderBox::applyVisualEffectOverflow const): Adjust the shadow offsets as necessary.
(WebCore::RenderBox::flippedClientBoxRect const): Drive-by cleanup.

Canonical link: https://commits.webkit.org/295977@main
